### PR TITLE
Add support for AWS EC2 Spot Instances

### PR DESCRIPTION
**Pull request by flfernandez**
AWS EC2 Spot instances can significantly reduce costs (compared to on-demand). By specifying the optional cluster definition variable `spot`, the EC2 can be created as part of a spot request. The example below has been tested successfully (t3.micro spot price in eu-west-1: U$0.0114):
```
cluster_vars:
  dev:
    hosttype_vars:
      node:
          auto_volumes:
          - device_name: "/dev/sda1"
            mountpoint: "/"
            fstype: "xfs"
            volume_type: "gp2"
            volume_size: 8
            encrypted: True
            delete_on_termination: True
          flavor: t3.micro
          spot:
            spot_price: 0.0114
            spot_launch_group: "{{ cluster_name }}"
            spot_wait_timeout: 600
            instance_initiated_shutdown_behavior: terminate
          vms_by_az:
              a: 1
              b: 1
              c: 1
```

The variable `spot` is optional, hence the omit in `get_cluster_hosts_target.yml`:
```
'spot': cluster_vars[buildenv].hosttype_vars[hostttype].spot|default(omit),
```
Ref: https://docs.ansible.com/ansible/latest/collections/amazon/aws/ec2_module.html

### CODE REVIEW

**Comment by dseeley-sky** [comment]
`instance_initiated_shutdown_behavior` is not related to `spot`, so probably shouldn't be in the same hierarchy.  It's useful, but probably belongs next to the other `instance_...` variables.

Strictly-speaking, you don't need to add this (or the `spot` variables) to `cluster_hosts_target` either, you can reference them directly from `cluster_vars` (`{{cluster_vars[buildenv].hosttype_vars[item.hosttype].spot_type | default(omit)}}` etc.).  What belongs in `cluster_hosts_target` is a bit inconsistent to be fair, but it is mostly used for variables that have to be derived from `cluster_vars` _per-instance_ (e.g. `hostname`, `hosttype`, `az_name`), or are mutated during the playbook run (`auto_volumes`).  (The exception is `flavor`, which should probably be removed.)

**Comment by dseeley-sky** [comment]
Otherwise - great addition!  

It might also be useful to put an example alongside the existing examples in https://github.com/sky-uk/clusterverse/blob/b93aba5d54497a502756458acb57f33c14b7b6c8/EXAMPLE/cluster_defs/aws/testid/eu-west-1/sandbox/cluster_vars__buildenv.yml#L34
Maybe something like (or whatever - this is what I tested it with):
```

      sysspot:
        auto_volumes: [ ]
        flavor: t3a.nano
        spot: { spot_price: 0.003, spot_launch_group: "{{ cluster_name }}", spot_wait_timeout: 600 }
        version: "{{sys_version | default('')}}"
        vms_by_az: { a: 1, b: 1, c: 0 }
```

Not sure where in the hierarchy `instance_initiated_shutdown_behavior: terminate` should live...?  Should it be per-hosttype (i.e. here, above), or per-cluster?

**Comment by flfernandez** [comment]
As per your suggestion I've moved spot properties out of get_cluster_hosts_target to be directly accessed from `{{cluster_vars[buildenv].hosttype_vars[item.hosttype].spot}}`
Also moved instance_initiated_shutdown_behavior one level up (out of `spot`).
Successfully tested creating as per below:

```
cluster_vars:
  dev:
    hosttype_vars:
      node:
          auto_volumes:
          - device_name: "/dev/sda1"
            mountpoint: "/"
            fstype: "xfs"
            volume_type: "gp2"
            volume_size: 10
            encrypted: True
            delete_on_termination: True
          - device_name: "/dev/sdb"
            mountpoint: "/data"
            fstype: "xfs"
            volume_type: "ephemeral"
            ephemeral: "ephemeral0"  
            lvmparams: { vg_name: "vg0", lv_name: "lv0", lv_size: "+100%FREE" }
          flavor: i3en.large
          spot:
            spot_price: 0.25
            spot_wait_timeout: 600
            spot_type: persistent
          instance_initiated_shutdown_behavior: terminate
          vms_by_az:
              a: 1
              b: 1
              c: 1
```

**Comment by dseeley-sky** [comment]
Redeploy issues:

> "Unable to stop instances: An error occurred (UnsupportedOperation) when calling the StopInstances operation: You can't stop the Spot Instance 'i-0dcca14c116dd377d' because it is associated with a one-time Spot Instance request. You can only stop Spot Instances associated with persistent Spot Instance requests."

**Comment by flfernandez** [comment]
> Redeploy issues:
> 
> > "Unable to stop instances: An error occurred (UnsupportedOperation) when calling the StopInstances operation: You can't stop the Spot Instance 'i-0dcca14c116dd377d' because it is associated with a one-time Spot Instance request. You can only stop Spot Instances associated with persistent Spot Instance requests."
> 
> What are the implications of a `persistent` spot type? Should they be `persistent` by default?

In order to stop a spot instance this option needs to be in place:
  spot_type: persistent

https://aws.amazon.com/about-aws/whats-new/2020/01/amazon-ec2-spot-instances-stopped-started-similar-to-on-demand-instances/

**Comment by dseeley-sky** [comment]
> In order to stop a spot instance this option needs to be in place:
> spot_type: persistent
> 
> https://aws.amazon.com/about-aws/whats-new/2020/01/amazon-ec2-spot-instances-stopped-started-similar-to-on-demand-instances/

It seems you can only stop an instance (e.g. for redeploy), if the spot _request_ is still in the **active** state.  By the defaults in the Ansible module, the stop request timeout is is is set to 10 minutes (https://docs.ansible.com/ansible/latest/collections/amazon/aws/ec2_module.html#parameter-spot_wait_timeout), after which it becomes **cancelled**, so you cannot stop, and the redeploy fails.  Interestingly, when you create a spot request using the AWS GUI, it sets the timeout to 1 year, so this problem disappears, (unless the request is cancelled by Amazon in the meantime).  

Basically, redeploy with spot instances either needs to be disabled, or we should set the default timeout to be a lot longer (say 1 week?) - thoughts?

**Comment by flfernandez** [comment]
> It seems you can only stop an instance (e.g. for redeploy), if the spot _request_ is still in the **active** state. By the defaults in the Ansible module, the stop request timeout is is is set to 10 minutes (https://docs.ansible.com/ansible/latest/collections/amazon/aws/ec2_module.html#parameter-spot_wait_timeout), after which it becomes **cancelled**, so you cannot stop, and the redeploy fails. Interestingly, when you create a spot request using the AWS GUI, it sets the timeout to 1 year, so this problem disappears, (unless the request is cancelled by Amazon in the meantime).
> 
> Basically, redeploy with spot instances either needs to be disabled, or we should set the default timeout to be a lot longer (say 1 week?) - thoughts?

My idea was that you can redeploy using spot instances **as long as** a rollback is not invoked. Which means as long as the instance is not attempted to be started back again.

In case of Cassandra/Scylla I'm particularly interested in the __scheme_addnewvm_rmdisk_rollback__ where instead of stopping the instance, it could be deleted.

Suggestion: Would it be a possible option to choose the desired state of an instance being redeployed? At the moment it is hard coded to *Stop* in the __scheme_addnewvm_rmdisk_rollback__.

**Comment by dseeley-sky** [comment]
> > It seems you can only stop an instance (e.g. for redeploy), if the spot _request_ is still in the **active** state. By the defaults in the Ansible module, the stop request timeout is is is set to 10 minutes (https://docs.ansible.com/ansible/latest/collections/amazon/aws/ec2_module.html#parameter-spot_wait_timeout), after which it becomes **cancelled**, so you cannot stop, and the redeploy fails. Interestingly, when you create a spot request using the AWS GUI, it sets the timeout to 1 year, so this problem disappears, (unless the request is cancelled by Amazon in the meantime).
> > Basically, redeploy with spot instances either needs to be disabled, or we should set the default timeout to be a lot longer (say 1 week?) - thoughts?
> 
> My idea was that you can redeploy using spot instances **as long as** a rollback is not invoked. Which means as long as the instance is not attempted to be started back again.
> 
> In case of Cassandra/Scylla I'm particularly interested in the **scheme_addnewvm_rmdisk_rollback** where instead of stopping the instance, it could be deleted.
> 
> Suggestion: Would it be a possible option to choose the desired state of an instance being redeployed? At the moment it is hard coded to _Stop_ in the **scheme_addnewvm_rmdisk_rollback**.

rollback can only work if the machine is stopped as opposed to terminated.

**Comment by flfernandez** [comment]
> rollback can only work if the machine is stopped as opposed to terminated.

Sure, sorry I wasn't clear before. I meant having a redeployment without rollback.

**Comment by dseeley-sky** [comment]
> > rollback can only work if the machine is stopped as opposed to terminated.
> 
> Sure, sorry I wasn't clear before. I meant having a redeployment without rollback.

Hmm - that would mean quite a big change to the redeploy scheme (scheme_addnewvm_rmdisk__**rollback**_)

**Comment by dseeley-sky** [comment]
The scheme **_scheme_rmvm_rmdisk_only** does not do rollback - this might work for you?

**Comment by flfernandez** [comment]
> The scheme **_scheme_rmvm_rmdisk_only** does not do rollback - this might work for you?

Not exactly. My idea for a Cassandra/Scylla redeploy where the data is on ephemeral disks is to: have both instances (new and old) running simultaneously and perform a rsync of the data (Predelete step); once completed the old instance is deleted and in the new one the service is started. At least in theory this can work. ;) 

Strictly speaking currently there is no redeploy scheme that would do that, but the scheme_addnewvm_rmdisk_rollback is the closest (without invoking rollback).

**Comment by dseeley-sky** [comment]
Elsewhere, the predelete step is often used to wait for the DB cluster to automatically rehydrate the new DB node from the rest of the cluster.  Achieves very similar to an rsync (which would require you to stop the DB on the node anyway).  If you want to do it using rsync, then I suggest we just set the spot expiry to be longer (otherwise we have to reengineer the redeploy scheme).  Is the motivation for using spot instances purely for testing (you're not planning to use spot instances in production)?  If so, then the duration only needs to be long enough to test the redeploy.

**Comment by flfernandez** [comment]
The idea is to use in CI jobs, development and tests (specially Testlabs) - so everything but production.
Which setting do you think would work for spot expiry?

**Comment by dseeley-sky** [comment]
`spot_wait_timeout` is the parameter to change.  If this is just for short-lived pipeline testing, maybe defaulting it to 3hrs?  Would that be enough?  I succeeded with 20m, but that was a single redeploy test.

**Review by dseeley-sky — APPROVED** [review]
